charms/slurmd/src/utils/rdma.py

```diff
@@ -0,0 +1,86 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Manage RDMA packages on compute nodes."""
+
+import logging
+
+import charms.operator_libs_linux.v0.apt as apt
+
+_logger = logging.getLogger(__name__)
+
+
+class RDMAOpsError(Exception):
+    """Exception raised when an RDMA package installation operation failed."""
+
+    @property
+    def message(self) -> str:
+        """Return message passed as argument to exception."""
+        return self.args[0]
+
+
+def install() -> None:
+    """Install RDMA packages.
+
+    Raises:
+        RDMAOpsError: Raised if error is encountered during package install.
+    """
+    install_packages = ["rdma-core", "infiniband-diags"]
+
+    _logger.info("installing RDMA packages: %s", install_packages)
+    try:
+        apt.add_package(install_packages)
+    except (apt.PackageNotFoundError, apt.PackageError) as e:
+        raise RDMAOpsError(f"failed to install packages {install_packages}. reason: {e}")
+
+    # Re-enable UCX/UCT.
+    # The OpenMPI package from archive includes a Debian patch that disables UCX to silence
+    # warnings when running on systems without RDMA hardware. This leads to the less performant
+    # "ob1" method being selected on Infiniband-enabled systems. By default, OpenMPI selects UCX
+    # when Infiniband devices are available. This default functionality is restored here.
+    #
+    # See:
+    # https://sources.debian.org/src/openmpi/4.1.4-3/debian/patches/no-warning-unused.patch/#L24
+    # https://github.com/open-mpi/ompi/issues/8367
+    # https://github.com/open-mpi/ompi/blob/v4.1.x/README#L763
+    path = "/etc/openmpi/openmpi-mca-params.conf"
+    _logger.info("enabling OpenMPI UCX transport in %s", path)
+
+    try:
+        with open(path, "r") as f:
+            lines = f.readlines()
+    except FileNotFoundError:
+        _logger.warn("%s not found. skipping OpenMPI UCX enablement", path)
+        return
+
+    # Current patch adds lines:
+    #   btl = ^uct,openib,ofi
+    #   pml = ^ucx
+    #   osc = ^ucx,pt2pt
+    # Remove "uct" and "ucx". Values must still begin with "^", e.g. `btl = ^openib,ofi`, to leave
+    # openib and ofi disabled.
+    new_lines = []
+    for line in lines:
+        if line.startswith(("btl =", "pml =", "osc =")):
+            parts = line.split()
+            values = parts[2].strip().lstrip("^").split(",")
+            values = [v for v in values if v not in ("uct", "ucx")]
+            # Remove line entirely if all values removed, e.g. "pml = ^ucx"
+            if values:
+                new_lines.append(f"{parts[0]} = ^{','.join(values)}\n")
+        else:
+            new_lines.append(line)
+
+    with open(path, "w") as f:
+        f.writelines(new_lines)
```

Comment:
Also, why not modify the file content in-place rather than create a duplicate list?

```suggestion
    file = Path("/etc/openmpi/openmpi-mca-params.conf")
    if not file.exists():
        _logger.warn("%s not found. skipping OpenMPI UCX enablement", file)
        return
        
    content = file.read_text().splitlines()
    for i, line in enumerate(content):
        if not line.startswith(("btl =", "pml =", "osc =")):
            continue
            
        parts = line.split()
        values = parts[2].strip().lstrip("^").split(",")
        values = [v for v in values if v not in ("uct", "ucx")]
        # Remove line entirely if all values removed, e.g. "pml = ^ucx"
        if values:
            content[i] = f"{parts[0]} = ^{','.join(values)}\n"
            
    file.write_text("\n".join(content) + "\n")
```

Figured this why we don't need to spend time expanding a new list in memory if we already have a fixed sized list we can create from _openmpi-mca-params.conf_. We also only need to worry about modifying the exact entries that we need to modify. Other benefit is one less level of nesting :sweat_smile:

Reply:
I've slightly modified the suggested change as giving it a conf file with:

```
mtl = ^ofi
btl = ^uct,openib,ofi
pml = ^ucx
osc = ^ucx,pt2pt
```

results in:

```
mtl = ^ofi
btl = ^openib,ofi

pml = ^ucx
osc = ^pt2pt
```

To handle cases like `pml = ^ucx`, I now set `content[i] = ""` when there are no `values` and filter out falsy entries in `content` as it's written.

I went with this over `del`-ing the entry (as we spoke about in yesterday's standup) to avoid the headaches associated with removing elements from a list as it's being iterated over. (Repeated `del` in a loop would also be expensive, however `openmpi-mca-params.conf` is going to be kilobytes at most so whichever method we go with should be fine).